### ISSUE TITLE
Showcase: Build an AI RAG Application for your Business | by Alexander Stahl | Medium

### ISSUE THREAD

**Issue by tekk** [tekk/webclips#3]
[

![](https://miro.medium.com/v2/resize:fill:88:88/1*wh8pStc929dLCM7gMjZxEw.jpeg)

](https://medium.com/@stahl950?source=post_page-----c354238b1dae--------------------------------)

![](https://miro.medium.com/v2/resize:fit:875/1*ehwpAx6tfk3D6EezcxPehw.png)

[https://www.linkedin.com/pulse/showcase-build-ai-rag-application-your-business-alexander-stahl-zp5mf/](https://www.linkedin.com/pulse/showcase-build-ai-rag-application-your-business-alexander-stahl-zp5mf/)

Since we never went into the technical implementation in the last 52 posts, I would like this to be the first post that shows how to build a practical AI solution.

Today again on the topic of RAG. Context-based AI solutions are the only way to build truly long-term relevant tools that create real added value without having to change AIaaS tools every few months. That’s why in this post I’ll show you how to build a simple RAG solution. I’ll start by briefly explaining the definition, how the whole thing works, how it differs from conventional LLM-generated content, how it is already being used in companies today — and then I’ll move on to the practical implementation.

Retrieval-augmented generation, or RAG, **combines knowledge from a curated knowledge base with language capabilities of an LLM to provide accurate and well-structured answers.** In RAG, we use a knowledge base for context-specific knowledge and an LLM for language generation, providing the best of both worlds. When a user asks a question through a prompt, the knowledge base provides contextual knowledge and the LLM provides well-structured answers.

**With RAG, we can use enterprise and confidential data sources to answer questions.** This is not possible when using third-party LLMs. It allows us to combine multiple data sources in different formats to create a knowledge base. We can use product manuals in PDF format, support tickets from a ticketing system, and content from web pages together in a single knowledge base. We can also curate the input data sources to extract only relevant knowledge and use it for RAG. This data can be continuously updated and pruned to keep it current.

**To find answers to queries, we can combine scalar and vector searches.** Vector searches can be used to find relevant answers in vectors, while scalar filters can help narrow down the context. For example, if the user asks a troubleshooting question about a specific product, scalar filters can be used to filter answers for that specific product.

RAG can use standard and out-of-the-box LLMs for language generation without creating or fine-tuning custom models, which saves a lot of money.

The differences between conventional LLM content creation and RAG solutions are easy to demonstrate.

**LLM Challenges:** Traditional LLMs can provide confident but outdated or unsourced answers, leading to potential inaccuracies and a lack of credibility.

![](https://miro.medium.com/v2/resize:fit:875/1*qV0lTwejoqZr_5eO6LhhCw.png)

[https://datasciencedojo.com/blog/rag-vs-finetuning-llm-debate/](https://datasciencedojo.com/blog/rag-vs-finetuning-llm-debate/)

**RAG Mechanism:** RAG addresses these issues by instructing the LLM to first retrieve relevant content from a specified content store (open or closed sources) before generating a response, ensuring the answer is grounded in up-to-date information.

**Benefits of RAG:** This approach mitigates problems of outdated information and lack of sources, as the model uses the most current data available and provides evidence for its answers, reducing the likelihood of hallucinations and misinformation.

There are a lot of real-life examples of RAG applications. These can be found on interesting [Twitter accounts](https://x.com/chiefaioffice), which constantly share new AI start-up solutions, or on YouTube channels.

**First, there are interactive chatbots** that businesses use to communicate with their customers. See “[Klarna AI assistant handles two-thirds of customer service chats in its first month](https://www.klarna.com/international/press/klarna-ai-assistant-handles-two-thirds-of-customer-service-chats-in-its-first-month/)”.

Chatbots are now more powerful and can use RAG to answer customer questions about products and help troubleshoot problems. RAG can help in automated responses to customer queries by email. Similar to chatbots, the responses can contain detailed information and answers to these queries.

RAG can help with root cause analysis of technical issues faced. Based on log messages, absorbed metrics, and information from manuals, RAG can help predict potential root causes quickly and aid in resolution of such issues in a timely manner.

Enterprises have help desk for functions like human resources, legal, or logistics. These functions can be automated with RAG to help employees find quick answers to their questions and problems. Enterprises have large document hubs with several hundred documents, and searching for specific information across these documents is laborious. RAG can help build document search capabilities on these hubs, and provide quick pointers to these documents that are relevant to the user search.

**Here are my top 3 RAG solutions from real start-ups:**

1.  [https://www.luminance.com/](https://www.luminance.com/)
2.  [https://www.ivo.ai/](https://www.ivo.ai/)
3.  [https://www.leya.law/](https://www.leya.law/)

A RAG system leverages both retrieval and generation capabilities to provide answers based on a combination of external and internal data sources. [Watch the video to get a perfect understanding.](https://www.youtube.com/watch?v=TRjq7t2Ms5I)

We use Next.js for the frontend. This RAG application will allow you to ingest various types of documents, create embeddings, store them in a vector database, retrieve relevant documents based on user queries, and generate responses using Mistral. **The provided examples are set up to work locally by default.**

Ensure you have the necessary tools and libraries installed. You will need:

*   Node.js and Next.js for the frontend
*   Python for the backend
*   Mistral for the language model
*   Pinecone as Vector database
*   PyMuPDF as a Document parser
*   BeautifulSoup for Web scraping (retrieving information from URLs)

```
npx create\-next\-app@latest rag-app  
cd rag-app  
npm install axios
```

The Python backend will handle document ingestion, embedding creation, retrieval, and response generation using [Mistral AI](https://www.linkedin.com/company/mistralai/).

```
pip install flask flask-cors requests pinecone-client transformers pymupdf beautifulsoup4
```

```
from flask import Flask, request, jsonify  
from flask\_cors import CORS  
import requests  
import pinecone  
import fitz    
from bs4 import BeautifulSoup  
from transformers import AutoTokenizer, AutoModel
```

```
app = Flask(\_\_name\_\_)  
CORS(app)\# Initialize Pinecone  
pinecone.init(api\_key="YOUR\_PINECONE\_API\_KEY", environment="us-west1-gcp")  
index\_name = "business-resources"  
if index\_name not in pinecone.list\_indexes():  
    pinecone.create\_index(index\_name, dimension=768)  
index = pinecone.Index(index\_name)\# Load transformer model  
tokenizer = AutoTokenizer.from\_pretrained("bert-base-uncased")  
model = AutoModel.from\_pretrained("bert-base-uncased")def create\_embedding(text):  
    inputs = tokenizer(text, return\_tensors="pt", max\_length=512, truncation=True)  
    outputs = model(\*\*inputs)  
    return outputs.last\_hidden\_state.mean(dim=1).detach().numpy()def add\_documents\_to\_pinecone(documents):  
    for doc\_id, doc in enumerate(documents):  
        embedding = create\_embedding(doc)  
        index.upsert(\[(str(doc\_id), embedding)\])def retrieve\_documents(query, top\_k=5):  
    query\_embedding = create\_embedding(query)  
    result = index.query(queries=query\_embedding.tolist(), top\_k=top\_k)  
    return \[int(match\['id'\]) for match in result\['matches'\]\]def generate\_response(query, documents):  
    context = " ".join(documents)  
    payload = {"context": context, "question": query}  
    response = requests.post('https://api.mistral.ai/generate', json=payload)  
    return response.json()\['answer'\]@app.route('/ask', methods=\['POST'\])  
def ask():  
    query = request.json.get('query')  
    doc\_ids = retrieve\_documents(query)  
    retrieved\_docs = \[documents\[doc\_id\] for doc\_id in doc\_ids\]  
    answer = generate\_response(query, retrieved\_docs)  
    return jsonify({"answer": answer})if \_\_name\_\_ == "\_\_main\_\_":  
    app.run(port=5000)
```

```
def ingest\_pdf(file\_path):  
    document = fitz.open(file\_path)  
    text = ""  
    for page\_num in range(len(document)):  
        page = document.load\_page(page\_num)  
        text += page.get\_text()  
    return text
```

```
def ingest\_word(file\_path):  
    from docx import Document  
    doc = Document(file\_path)  
    text = ""  
    for para in doc.paragraphs:  
        text += para.text  
    return textdef ingest\_web\_page(url):  
    response = requests.get(url)  
    soup = BeautifulSoup(response.content, 'html.parser')  
    return soup.get\_text()\# Example usage  
documents = \[  
    ingest\_pdf('path/to/document.pdf'),  
    ingest\_word('path/to/document.docx'),  
    ingest\_web\_page('https://example.com')  
\]  
add\_documents\_to\_pinecone(documents)
```

```
  
import { NextResponse } from 'next/server';  
import axios from 'axios';
```

```
export async function POST(request) {  
  const { query } = await request.json();  
  try {  
    const response = await axios.post('http://localhost:5000/ask', { query });  
    return NextResponse.json(response.data);  
  } catch (error) {  
    return NextResponse.json({ error: 'Failed to get answer' }, { status: 500 });  
  }  
}
```

```
// app/page.js  
'use client';
```

```
import { useState } from 'react';export default function Home() {  
  const \[query, setQuery\] = useState('');  
  const \[answer, setAnswer\] = useState(''); const handleSubmit = async (e) => {  
    e.preventDefault();  
    const res = await fetch('/api/ask', {  
      method: 'POST',  
      headers: {  
        'Content-Type': 'application/json',  
      },  
      body: JSON.stringify({ query }),  
    });  
    const data = await res.json();  
    setAnswer(data.answer);  
  }; return (  
    <div>  
      <h1>Ask a Question</h1>  
      <form onSubmit={handleSubmit}>  
        <input   
          type="text"   
          value={query}   
          onChange={(e) => setQuery(e.target.value)}   
          placeholder="Enter your question"  
        />  
        <button type="submit">Ask</button>  
      </form>  
      <div>  
        <h2>Answer:</h2>  
        <p>{answer}</p>  
      </div>  
    </div>  
  );  
}
```

```
python app.py  
npm run dev
```

This setup uses Next.js for the front end and a Python backend to handle document ingestion, embedding creation, retrieval, and interaction with Mistral for response generation. This approach provides a seamless and interactive user experience for querying and obtaining information from your business resources.

This way you can easily communicate with your business information and make better decisions. Would you like to get access to this application? Then send me a DM.

The [best way to get started](https://www.simpleai.at/) in your own company is by analyzing your potential for AI-driven solutions based on your data.

![](https://miro.medium.com/v2/resize:fit:875/0*rOTiKXPv000v1P4r)

I’m offering business owners a 30-minute consultation to understand how AI-based solutions can lead to more effectiveness, cost savings, and profit maximization in their industry.

Take advantage of this opportunity and book a consultation with [Simple AI](https://www.linkedin.com/company/simpleaiae/): [https://calendly.com/alexanderstahl/30min?back=1&month=2023-10](https://calendly.com/alexanderstahl/30min?back=1&month=2023-10)

Thank you for taking the time to read this article and I look forward to your feedback.
